Reinit systems once

roystgnr opened the pull request:
This PR starts with a (small) libMesh update, required to get access to the new EquationSystems::reinit_solutions() and EquationSystems::reinit_systems() APIs there.

Other changes since the last (recent!) libMesh update:

* Fixed corner case in PetscVector<Complex>::localize_to_one
* New infinite element example miscellaneous_ex14
* DistributedMesh infinite element fixes

Then, FEProblemBase::meshChanged() is refactored to optionally allow a partial reinit using those APIs, and FEProblemBase::adaptMesh() uses those partial reinit capabilities to shave a few percent off of runtime (Refs #10256) in applications which do multiple adaptivity cycles per solve step, as discussed in https://github.com/libMesh/libmesh/pull/1497 and [on moose-users](https://groups.google.com/forum/#!msg/moose-users/ewSE3PsOkoE/FRIAu3P7CwAJ)

roystgnr committed:
Comment FEProblemBase::meshChanged()

Now the operation of this class is 0.01% less confusing.

roystgnr committed:
Add FEProblemBase::meshChangedHelper()

This allows us to provide an optional boolean argument which signifies
an "intermediate change": a mesh change which will soon be followed by
another mesh change and which therefore does not require systems to be
reinitialized on the intermediate mesh.

roystgnr committed:
Intermediate meshChangedHelper calls in adaptMesh

This skips unnecessary Jacobian sparsity recalculation and memory
reallocation in intermediate phases of multi-step mesh adaptivity,
which speeds up such applications by a couple percent.

Refs #10256

moosebuild commented:
Job [Precheck](https://www.moosebuild.org/job/134273/) on cef2ec5 wanted to post the following:

Your code requires style changes.
In the top level of your repository, you can run:

`git clang-format 08bb17a1b082a440272069e3332241e8c225f840`

Alternatively, apply the patch found [here](http://mooseframework.org/docs/PRs/10257/style.patch).
You can directly apply the patch by running, in the top level of your repository:

`curl -s http://mooseframework.org/docs/PRs/10257/style.patch | git apply -v`

moosebuild commented on libmesh:
```diff
@@ -1 +1 @@
-Subproject commit 7f3ced2710c67ef3f7d7ba512090259b09997c74
+Subproject commit 8adfd6bdfb08f12abe6de0102d16c4c6a6632bf1
```
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

roystgnr committed:
Satisfy clang-format

Really, there's no Excuse for two Spaces In between a Statement and
The following comment.  Those of You who disagree should Relent And
Not Try to Squabble.

jessecarterMOOSE commented:
I'm also seeing a few (single-digit) percent speedups with this, depending on the problem.

jessecarterMOOSE commented:
Hmm, now wait just a minute. I see some speed up on a small problem based on this patch, but some of my bigger problems are running _slightly slower_.

I'm testing on falcon and run-to-run variability is generally less than a tenth of a percent, (which is a lot better than our systems!). 

I ran my PRARIEDOG mesh adaptivity benchmark problem in 2D up through a 200x200 element initial mesh. First I ran with a recent MOOSE branch, then updated libmesh to (yesterday's) master (which contains libMesh/libmesh#1497), and ran again. Runtimes were more or less the same, definitely less than a percent (up or down) in all cases, which was more than I expected, but I moved on...

Then I applied this patch and re-ran and compared times to the previous run. The results started coming in for the smaller problems first so I made my comment above, but then the bigger problems came in... I plotted the results below. Lower cpu counts show more speedup in the small problems but also show more slowdown at the bigger problems. You can find the data [here](https://docs.google.com/spreadsheets/d/1yq6OYybQYKdT9qhHvK_9L6KiZAYygzgbRruQ0gcRUuw/edit?usp=sharing). Is this real or some artifact of the way I'm testing??

I'm also curious what the new benchmarking system has to say about this...

![canvas](https://user-images.githubusercontent.com/12945054/32913506-8b43b6d6-cae0-11e7-9061-79ca6fcfa292.png)

roystgnr commented:
I definitely can't explain any of that.  This patch should do nothing except *remove* redundant calls to system reinit().  I don't see how doing that could slow down anything!

jessecarterMOOSE commented:
That's what I thought too! I'll have to dig some more to see where all that time is going.

Another thing to think about is how my problem operates. When it's time to refine or coarsen, it generally performs several (3-6+) cycles of adaptivity. That means that previously, it would do, say, 6 "full" (solutions + system) reinits and be done. Now it would do 6 solution reinits and then 1 system reinit after it exists the cycling loop. Could there be any benefit to doing a system reinit immediately following a relatively "small" solution reinit, as opposed to what it's doing now by trying to reinit a system that's "way behind"?

roystgnr commented:
> Could there be any benefit to doing a system reinit immediately following a relatively "small" solution reinit, as opposed to what it's doing now by trying to reinit a system that's "way behind"?

Not that I can imagine, although you're going down the same track that I tried to.  Regardless of how big the change is, we end up clearing matrices, recomputing sparsity from scratch, and reallocating matrices whole.

jessecarterMOOSE commented:
@roystgnr For completeness you might want to add the same logic to [FEProblemBase::initialAdaptMesh()](https://github.com/idaholab/moose/blob/devel/framework/src/base/FEProblemBase.C#L4430). With your patch currently, the full `_eq.reinit()` is getting called following each cycle of initial mesh adaptivity. Of course initial adaptivity is only done once, but the first refinement could be particularly heavy.

rwcarlsen commented:
Unless you've done work to properly isolate and quiet your system, I wouldn't trust any benchmark numbers - especially without some statistical analysis over several runs.  I have also found that code/data alignment shifts in moose executables from benign code changes can cause +/-  ~1% changes to performance.

jessecarterMOOSE commented:
I've been running things on falcon where the run-to-run variability is about 0.1% or less. In each case I usually run the problem 3 times and take the minimum. Still though, I agree there is definitely some randomness to account for. Perhaps I need to run this through that benchmarking system I keep hearing about, eh @rwcarlsen ?

rwcarlsen commented:
There is some info in ``[moose]/scripts/README.md`` on how to manually run your own benchmarks.

jessecarterMOOSE commented:
Cool beans, @rwcarlsen 

Is there any chance of "isolating" processors on something you don't have root access to, like falcon?

rwcarlsen commented:
No.  Just make sure there is as little running on the machine as possible.

On Wed, Nov 22, 2017 at 4:23 PM, Jesse Carter <notifications@github.com>
wrote:

> Cool beans, @rwcarlsen <https://github.com/rwcarlsen>
>
> Is there any chance of "isolating" processors on something you don't have
> root access to, like falcon?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/pull/10257#issuecomment-346499278>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ABSeSjdnXrAUDcsRrSjkJutwXeZZMw_fks5s5KzVgaJpZM4QfF6B>
> .
>
